pkg/querier/queryrange/querysharding.go

```diff
@@ -0,0 +1,246 @@
+package queryrange
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/loki/pkg/loghttp"
+	"github.com/grafana/loki/pkg/logql"
+	"github.com/grafana/loki/pkg/logql/marshal"
+)
+
+var nanosecondsInMillisecond = int64(time.Millisecond / time.Nanosecond)
+
+// NewQueryShardMiddleware creates a middleware which downstreams queries after AST mapping and query encoding.
+func NewQueryShardMiddleware(
+	logger log.Logger,
+	confs queryrange.ShardingConfigs,
+	minShardingLookback time.Duration,
+	middlewareMetrics *queryrange.InstrumentMiddlewareMetrics,
+	shardingMetrics *logql.ShardingMetrics,
+) queryrange.Middleware {
+
+	noshards := !hasShards(confs)
+
+	if noshards {
+		level.Warn(logger).Log(
+			"middleware", "QueryShard",
+			"msg", "no configuration with shard found",
+			"confs", fmt.Sprintf("%+v", confs),
+		)
+		return queryrange.PassthroughMiddleware
+	}
+
+	mapperware := queryrange.MiddlewareFunc(func(next queryrange.Handler) queryrange.Handler {
+		return newASTMapperware(confs, next, logger, shardingMetrics)
+	})
+
+	return queryrange.MiddlewareFunc(func(next queryrange.Handler) queryrange.Handler {
+		return &shardSplitter{
+			MinShardingLookback: minShardingLookback,
+			shardingware: queryrange.MergeMiddlewares(
+				queryrange.InstrumentMiddleware("shardingware", middlewareMetrics),
+				mapperware,
+			).Wrap(next),
+			now:  time.Now,
+			next: queryrange.InstrumentMiddleware("sharding-bypass", middlewareMetrics).Wrap(next),
+		}
+	})
+
+}
+
+func newASTMapperware(
+	confs queryrange.ShardingConfigs,
+	next queryrange.Handler,
+	logger log.Logger,
+	metrics *logql.ShardingMetrics,
+) *astMapperware {
+
+	return &astMapperware{
+		confs:   confs,
+		logger:  log.With(logger, "middleware", "QueryShard.astMapperware"),
+		next:    next,
```

Comment:
`next` is a confusing name, at first I assumed this was used for Downstream queries.

Maybe `fallback` ?

Reply:
My bad got it now !